File: src/tools/add-field.ts
import { readFileSync, writeFileSync } from 'fs';
import { join } from 'path';

export interface AddFieldParams {
  resource: string;
  field_name: string;
  field_type: 'string' | 'int64' | 'bool' | 'object' | 'array';
  proto_path: string;
  parent_type?: string;
  description?: string;
  json_name?: string;
}

export function addField(
  repoPath: string,
  typesFile: string,
  params: AddFieldParams
): string {
  const filePath = join(repoPath, typesFile);
  let content = readFileSync(filePath, 'utf-8');

  // Determine Go type
  let goType: string;
  switch (params.field_type) {
    case 'string':
      goType = '*string';
      break;
    case 'int64':
      goType = '*int64';
      break;
    case 'bool':
      goType = '*bool';
      break;
    case 'array':
      goType = '[]string'; // Default, can be customized
      break;
    case 'object':
      goType = `*${params.resource}_${params.field_name}`;
      break;
    default:
      throw new Error(`Unsupported field type: ${params.field_type}`);
  }

  // JSON name (default to camelCase field name)
  const jsonName = params.json_name || toCamelCase(params.field_name);

  // Build field definition
  const fieldDef = buildFieldDefinition(
    params.field_name,
    goType,
    jsonName,
    params.proto_path,
    params.description
  );

  // Find insertion point
  const parentType = params.parent_type || `${params.resource}Spec`;
  const insertionPoint = findInsertionPoint(content, parentType);

  if (insertionPoint === -1) {
    throw new Error(
      `Could not find parent type: ${parentType}\n\n` +
        `Make sure the type exists in ${typesFile}`
    );
  }

  // Insert field
  const lines = content.split('\n');
  lines.splice(insertionPoint, 0, fieldDef);
  content = lines.join('\n');

  // Write back
  writeFileSync(filePath, content, 'utf-8');

  return `✅ Added field to ${typesFile}\n\n${fieldDef}`;
}

function buildFieldDefinition(
  fieldName: string,
  goType: string,
  jsonName: string,
  protoPath: string,
  description?: string
): string {
  const lines: string[] = [];

  // Add description comment if provided
  if (description) {
    lines.push(`\t// ${description}`);
  }

  // Add proto annotation
  lines.push(`\t// +kcc:proto=${protoPath}`);

  // Add field
  lines.push(`\t${fieldName} ${goType} \`json:"${jsonName},omitempty"\``);

  return lines.join('\n');
}

function findInsertionPoint(content: string, parentType: string): number {
  const lines = content.split('\n');

  // Find the parent type struct
  let inStruct = false;
  let lastFieldLine = -1;

  for (let i = 0; i < lines.length; i++) {
    const line = lines[i];

    // Found the struct definition
    if (line.includes(`type ${parentType} struct {`)) {
      inStruct = true;
      continue;
    }

    // Inside the struct, track last field line
    if (inStruct) {
      // Check if this is a field line (contains backticks for json tag)
      if (line.includes('`json:')) {
        lastFieldLine = i;
      }

      // End of struct
      if (line.trim() === '}') {
        // Insert before the closing brace
        return lastFieldLine + 1;
      }
    }
  }

  return -1;
}

function toCamelCase(str: string): string {
  return str.charAt(0).toLowerCase() + str.slice(1);
}

File: go/internal/tools/add_field.go
package tools

import (
	"fmt"
	"os"
	"path/filepath"
	"strings"
)

// AddFieldParams contains parameters for adding a field
type AddFieldParams struct {
	Resource    string `json:"resource"`
	FieldName   string `json:"field_name"`
	FieldType   string `json:"field_type"` // "string", "int64", "bool", "object", "array"
	ProtoPath   string `json:"proto_path"`
	ParentType  string `json:"parent_type,omitempty"`
	Description string `json:"description,omitempty"`
	JSONName    string `json:"json_name,omitempty"`
}

// AddField adds a field to a KCC resource types file
func AddField(repoPath, typesFile string, params AddFieldParams) (string, error) {
	filePath := filepath.Join(repoPath, typesFile)

	// Read file
	contentBytes, err := os.ReadFile(filePath)
	if err != nil {
		return "", fmt.Errorf("failed to read file: %w", err)
	}
	content := string(contentBytes)

	// Determine Go type
	goType, err := getGoType(params.FieldType, params.Resource, params.FieldName)
	if err != nil {
		return "", err
	}

	// JSON name (default to camelCase field name)
	jsonName := params.JSONName
	if jsonName == "" {
		jsonName = toCamelCase(params.FieldName)
	}

	// Build field definition
	fieldDef := buildFieldDefinition(params.FieldName, goType, jsonName, params.ProtoPath, params.Description)

	// Find insertion point
	parentType := params.ParentType
	if parentType == "" {
		parentType = fmt.Sprintf("%sSpec", params.Resource)
	}

	insertionPoint := findInsertionPoint(content, parentType)
	if insertionPoint == -1 {
		return "", fmt.Errorf("could not find parent type: %s\n\nMake sure the type exists in %s", parentType, typesFile)
	}

	// Insert field
	lines := strings.Split(content, "\n")
	newLines := make([]string, 0, len(lines)+3)
	newLines = append(newLines, lines[:insertionPoint]...)
	newLines = append(newLines, fieldDef)
	newLines = append(newLines, lines[insertionPoint:]...)
	content = strings.Join(newLines, "\n")

	// Write back
	if err := os.WriteFile(filePath, []byte(content), 0644); err != nil {
		return "", fmt.Errorf("failed to write file: %w", err)
	}

	return fmt.Sprintf("✅ Added field to %s\n\n%s", typesFile, fieldDef), nil
}

// getGoType determines the Go type for a field
func getGoType(fieldType, resource, fieldName string) (string, error) {
	switch fieldType {
	case "string":
		return "*string", nil
	case "int64":
		return "*int64", nil
	case "bool":
		return "*bool", nil
	case "array":
		return "[]string", nil // Default, can be customized
	case "object":
		return fmt.Sprintf("*%s_%s", resource, fieldName), nil
	default:
		return "", fmt.Errorf("unsupported field type: %s", fieldType)
	}
}

// buildFieldDefinition builds the field definition with annotations
func buildFieldDefinition(fieldName, goType, jsonName, protoPath, description string) string {
	var lines []string

	// Add description comment if provided
	if description != "" {
		lines = append(lines, fmt.Sprintf("\t// %s", description))
	}

	// Add proto annotation
	lines = append(lines, fmt.Sprintf("\t// +kcc:proto=%s", protoPath))

	// Add field
	lines = append(lines, fmt.Sprintf("\t%s %s `json:\"%s,omitempty\"`", fieldName, goType, jsonName))

	return strings.Join(lines, "\n")
}

// findInsertionPoint finds where to insert the new field
func findInsertionPoint(content, parentType string) int {
	lines := strings.Split(content, "\n")

	// Find the parent type struct
	inStruct := false
	lastFieldLine := -1

	for i, line := range lines {
		// Found the struct definition
		if strings.Contains(line, fmt.Sprintf("type %s struct {", parentType)) {
			inStruct = true
			continue
		}

		// Inside the struct, track last field line
		if inStruct {
			// Check if this is a field line (contains backticks for json tag)
			if strings.Contains(line, "`json:") {
				lastFieldLine = i
			}

			// End of struct
			if strings.TrimSpace(line) == "}" {
				// Insert before the closing brace
				return lastFieldLine + 1
			}
		}
	}

	return -1
}

// toCamelCase converts PascalCase to camelCase
func toCamelCase(str string) string {
	if len(str) == 0 {
		return str
	}
	return strings.ToLower(str[:1]) + str[1:]
}
